File: building_blocks/union_find.py

class UnionFind:
    def __init__(self, size):
        self.ids = [i for i in range(size)]
        self.sizes = [1 for i in range(size)]

    def root(self, a):
        root = a
        while self.ids[root] != root:
            root = self.ids[root]

        # Path compression
        while a != root:
            t = self.ids[a]
            self.ids[a] = root
            a = t

        return root

    def find(self, a, b):
        return self.root(a) == self.root(b)

    def union(self, a, b):
        root_a = self.root(a)
        root_b = self.root(b)

        if self.sizes[root_a] >= self.sizes[root_b]:
            # merge b into a:
            self.ids[root_b] = root_a
            self.sizes[root_a] += self.sizes[root_b]
        else:
            self.ids[root_a] = root_b
            self.sizes[root_b] += self.sizes[root_a]


if __name__ == '__main__':
    print("Hello world")
    uf = UnionFind(10)
    uf.union(3, 4)
    print(uf.ids)

File: problems/src/main/java/common/UnionFind.java
package common;

public class UnionFind {
  private int[] ids;
  private int[] sizes;
  public UnionFind(int size) {
    sizes = new int[size];
    ids = new int[size];
    for (int i = 0; i < size; i++) {
      ids[i] = i;
      sizes[i] = 1;
    }
  }

  public int root(int i) {
    int root = i;
    while (root != ids[root]) {
      root = ids[root];
    }

    // Path compression:
    while (i != root) {
      int tmp = ids[i];
      ids[i] = root;
      i = tmp;
    }
    return root;
  }

  public boolean find(int a, int b) {
    return root(a) == root(b);
  }

  public void union(int a, int b) {
    int rootA = root(a);
    int rootB = root(b);

    // Balancing the tree:
    if (sizes[rootA] > sizes[rootB]) {
      ids[rootB] = rootA;
      sizes[rootA] += sizes[rootB];
    } else {
      ids[rootA] = rootB;
      sizes[rootB] += sizes[rootA];
    }
  }
}
